# Machine Learning algorithms

### Bioinfo group_assign_3

Import statements

In [1]:
import pandas as pd
from sklearn.model_selection import StratifiedKFold, cross_val_score
from sklearn import svm
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB

Load data

In [2]:
### Full dataframe
data=pd.read_csv("df.csv")
print(data)
### Sets for classification
X = data.iloc[:, :-1]
y = data['protFamily']
print(y)
print(X)

            AA        AR        AN        AD        AC        AQ        AE  \
0     0.014414  0.007207  0.000000  0.007207  0.007207  0.007207  0.000000   
1     0.000000  0.002920  0.011679  0.002920  0.002920  0.000000  0.002920   
2     0.000000  0.000000  0.000000  0.000000  0.000000  0.003591  0.000000   
3     0.000000  0.000000  0.000000  0.000000  0.000000  0.004301  0.000000   
4     0.006502  0.002890  0.001445  0.003251  0.001084  0.005418  0.003973   
...        ...       ...       ...       ...       ...       ...       ...   
1440  0.026667  0.000000  0.000000  0.026667  0.000000  0.013333  0.026667   
1441  0.033333  0.011111  0.000000  0.000000  0.000000  0.022222  0.011111   
1442  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
1443  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
1444  0.000000  0.000000  0.000000  0.000000  0.000000  0.012987  0.012987   

            AG        AH        AI  ...        VK        VM    

## Make and evaluate ML models

### SVM

In [3]:
### Make model and apply 10-fold cross-validation
svm_clf = svm.SVC(kernel='linear', C=1, random_state=42)
cross_val = StratifiedKFold(n_splits=10, shuffle=True, random_state=42)
svm_prec_scores = cross_val_score(svm_clf, X, y, cv=cross_val, scoring='precision_weighted')
svm_rec_scores = cross_val_score(svm_clf, X, y, cv=cross_val, scoring='recall_weighted')
svm_f1_scores = cross_val_score(svm_clf, X, y, cv=cross_val, scoring='f1_weighted')
### Get metrics
print("SVM classifier metrics across the 10 folds:")
print("| (Shown as ==> Avg +- StdDev)")
svm_mean_prec = svm_prec_scores.mean()
svm_prec_stdev = svm_prec_scores.std()
print(f"| Precision: {svm_mean_prec:.3f} +- {svm_prec_stdev:.3f}")
svm_mean_rec = svm_rec_scores.mean()
svm_rec_stdev = svm_rec_scores.std()
print(f"| Recall: {svm_mean_rec:.3f} +- {svm_rec_stdev:.3f}")
svm_mean_f1 = svm_f1_scores.mean()
svm_f1_stdev = svm_f1_scores.std()
print(f"| F1-score: {svm_mean_f1:.3f} +- {svm_f1_stdev:.3f}")

C:\Users\nunoc\AppData\Roaming\Python\Python311\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\nunoc\AppData\Roaming\Python\Python311\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\nunoc\AppData\Roaming\Python\Python311\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\nunoc\AppData\Roaming\Python

SVM classifier metrics across the 10 folds:
| (Shown as ==> Avg +- StdDev)
| Precision: 0.646 +- 0.005
| Recall: 0.803 +- 0.003
| F1-score: 0.716 +- 0.004


### Random Forests

In [4]:
### Make model and apply 10-fold cross-validation
rf_clf = RandomForestClassifier(random_state=42)
rf_prec_scores = cross_val_score(rf_clf, X, y, cv=cross_val, scoring='precision_weighted')
rf_rec_scores = cross_val_score(rf_clf, X, y, cv=cross_val, scoring='recall_weighted')
rf_f1_scores = cross_val_score(rf_clf, X, y, cv=cross_val, scoring='f1_weighted')
### Get metrics
print("Random Forests classifier metrics across the 10 folds:")
print("| (Shown as ==> Avg +- StdDev)")
rf_mean_prec = rf_prec_scores.mean()
rf_prec_stdev = rf_prec_scores.std()
print(f"| Precision: {rf_mean_prec:.3f} +- {rf_prec_stdev:.3f}")
rf_mean_rec = rf_rec_scores.mean()
rf_rec_stdev = rf_rec_scores.std()
print(f"| Recall: {rf_mean_rec:.3f} +- {rf_rec_stdev:.3f}")
rf_mean_f1 = rf_f1_scores.mean()
rf_f1_stdev = rf_f1_scores.std()
print(f"| F1-score: {rf_mean_f1:.3f} +- {rf_f1_stdev:.3f}")

Random Forests classifier metrics across the 10 folds:
| (Shown as ==> Avg +- StdDev)
| Precision: 0.991 +- 0.007
| Recall: 0.990 +- 0.007
| F1-score: 0.990 +- 0.007


### Naive Bayes

In [5]:
### Make model and apply 10-fold cross-validation
gnb = GaussianNB()
gnb_prec_scores = cross_val_score(gnb, X, y, cv=cross_val, scoring='precision_weighted')
gnb_rec_scores = cross_val_score(gnb, X, y, cv=cross_val, scoring='recall_weighted')
gnb_f1_scores = cross_val_score(gnb, X, y, cv=cross_val, scoring='f1_weighted')
### Get metrics
print("Gaussian naive Bayes classifier metrics across the 10 folds:")
print("| (Shown as ==> Avg +- StdDev)")
gnb_mean_prec = gnb_prec_scores.mean()
gnb_prec_stdev = gnb_prec_scores.std()
print(f"| Precision: {gnb_mean_prec:.3f} +- {gnb_prec_stdev:.3f}")
gnb_mean_rec = gnb_rec_scores.mean()
gnb_rec_stdev = gnb_rec_scores.std()
print(f"| Recall: {gnb_mean_rec:.3f} +- {gnb_rec_stdev:.3f}")
gnb_mean_f1 = gnb_f1_scores.mean()
gnb_f1_stdev = gnb_f1_scores.std()
print(f"| F1-score: {gnb_mean_f1:.3f} +- {gnb_f1_stdev:.3f}")

Gaussian naive Bayes classifier metrics across the 10 folds:
| (Shown as ==> Avg +- StdDev)
| Precision: 0.973 +- 0.005
| Recall: 0.969 +- 0.006
| F1-score: 0.970 +- 0.006


### Summary Table

-> Results after running the "classify_proteins.py" script (same as obtained here // best method in bold):

(Metrics shown as "Avg +- StDev")

| ML Algorithm | Precision | Recall | F1-score |
|:------------:|:---------:|:------:|:--------:|
| SVM | 0.646 +- 0.005 | 0.803 +- 0.003 | 0.716 +- 0.004 |
| **Random Forests** | **0.991 +- 0.007** | **0.990 +- 0.007** | **0.990 +- 0.007** |
| Gaussian naive Bayes | 0.973 +- 0.005 | 0.969 +- 0.006 | 0.970 +- 0.006 |